File: tests_opossum/Opossum.IntegrationTests/DecisionModel/BuildDecisionModelIntegrationTests.cs
using Opossum.Core;
using Opossum.DecisionModel;
using Opossum.DependencyInjection;
using Opossum.Exceptions;
using Opossum.Extensions;

namespace Opossum.IntegrationTests.DecisionModel;

/// <summary>
/// Integration tests for <see cref="DecisionModelExtensions.BuildDecisionModelAsync{TState}"/>.
/// Each test uses a completely isolated temp directory — no shared state, no collection needed.
/// </summary>
public sealed class BuildDecisionModelIntegrationTests : IDisposable
{
    private readonly string _baseTempPath;
    private readonly List<ServiceProvider> _serviceProviders = [];

    public BuildDecisionModelIntegrationTests()
    {
        _baseTempPath = Path.Combine(
            Path.GetTempPath(),
            "OpossumDecisionModelIntegrationTests",
            Guid.NewGuid().ToString());
    }

    public void Dispose()
    {
        foreach (var sp in _serviceProviders)
            sp.Dispose();

        if (Directory.Exists(_baseTempPath))
        {
            try
            { Directory.Delete(_baseTempPath, recursive: true); }
            catch { /* ignore cleanup errors */ }
        }
    }

    private IEventStore CreateEventStore()
    {
        var path = Path.Combine(_baseTempPath, Guid.NewGuid().ToString());
        var services = new ServiceCollection();
        services.AddOpossum(opt =>
        {
            opt.RootPath = path;
            opt.FlushEventsImmediately = false;
            opt.UseStore("TestContext");
        });
        var sp = services.BuildServiceProvider();
        _serviceProviders.Add(sp);
        return sp.GetRequiredService<IEventStore>();
    }

    #region Test domain events

    private record CourseCreatedEvent(Guid CourseId, int MaxStudents) : IEvent;
    private record StudentEnrolledEvent(Guid CourseId, Guid StudentId) : IEvent;
    private record UnrelatedEvent(string Data) : IEvent;

    #endregion

    #region Helper projections

    private static IDecisionProjection<bool> CourseExistsProjection(Guid courseId) =>
        new DecisionProjection<bool>(
            initialState: false,
            query: Query.FromItems(new QueryItem
            {
                EventTypes = [nameof(CourseCreatedEvent)],
                Tags = [new Tag("courseId", courseId.ToString())]
            }),
            apply: (_, evt) => evt.Event.Event is CourseCreatedEvent);

    private static IDecisionProjection<int> EnrollmentCountProjection(Guid courseId) =>
        new DecisionProjection<int>(
            initialState: 0,
            query: Query.FromItems(new QueryItem
            {
                EventTypes = [nameof(StudentEnrolledEvent)],
                Tags = [new Tag("courseId", courseId.ToString())]
            }),
            apply: (count, evt) => evt.Event.Event is StudentEnrolledEvent ? count + 1 : count);

    #endregion

    #region Empty store

    [Fact]
    public async Task BuildDecisionModelAsync_EmptyStore_StateIsInitialStateAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        var model = await store.BuildDecisionModelAsync(CourseExistsProjection(courseId));

        Assert.False(model.State);
    }

    [Fact]
    public async Task BuildDecisionModelAsync_EmptyStore_AfterSequencePositionIsNullAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        var model = await store.BuildDecisionModelAsync(CourseExistsProjection(courseId));

        Assert.Null(model.AppendCondition.AfterSequencePosition);
    }

    [Fact]
    public async Task BuildDecisionModelAsync_EmptyStore_FailIfEventsMatchIsProjectionQueryAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();
        var projection = CourseExistsProjection(courseId);

        var model = await store.BuildDecisionModelAsync(projection);

        Assert.Same(projection.Query, model.AppendCondition.FailIfEventsMatch);
    }

    #endregion

    #region State reflects appended events

    [Fact]
    public async Task BuildDecisionModelAsync_AfterAppend_StateReflectsEventsAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var model = await store.BuildDecisionModelAsync(CourseExistsProjection(courseId));

        Assert.True(model.State);
    }

    [Fact]
    public async Task BuildDecisionModelAsync_AfterMultipleEnrollments_CountIsCorrectAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        for (var i = 0; i < 3; i++)
        {
            await store.AppendAsync(
                new StudentEnrolledEvent(courseId, Guid.NewGuid())
                    .ToDomainEvent()
                    .WithTag("courseId", courseId.ToString()));
        }

        var model = await store.BuildDecisionModelAsync(EnrollmentCountProjection(courseId));

        Assert.Equal(3, model.State);
    }

    [Fact]
    public async Task BuildDecisionModelAsync_AfterSequencePosition_IsMaxEventPositionAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        await store.AppendAsync(
            new StudentEnrolledEvent(courseId, Guid.NewGuid())
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var model = await store.BuildDecisionModelAsync(EnrollmentCountProjection(courseId));

        Assert.Equal(2, model.AppendCondition.AfterSequencePosition);
    }

    [Fact]
    public async Task BuildDecisionModelAsync_UnrelatedEventsDoNotAffectStateAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();
        var otherCourseId = Guid.NewGuid();

        // Append events for a different course — should be invisible to our projection
        await store.AppendAsync(
            new CourseCreatedEvent(otherCourseId, 10)
                .ToDomainEvent()
                .WithTag("courseId", otherCourseId.ToString()));

        var model = await store.BuildDecisionModelAsync(CourseExistsProjection(courseId));

        Assert.False(model.State);
        Assert.Null(model.AppendCondition.AfterSequencePosition);
    }

    #endregion

    #region DCB round-trip: AppendCondition enforces consistency

    [Fact]
    public async Task BuildDecisionModelAsync_StaleCondition_AppendThrowsAsync()
    {
        // Arrange: append initial event, build decision model
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 1)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var staleModel = await store.BuildDecisionModelAsync(EnrollmentCountProjection(courseId));

        // Simulate concurrent write: another client enrols a student
        await store.AppendAsync(
            new StudentEnrolledEvent(courseId, Guid.NewGuid())
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        // Act & Assert: our stale AppendCondition must be rejected
        var newEvent = new StudentEnrolledEvent(courseId, Guid.NewGuid())
            .ToDomainEvent()
            .WithTag("courseId", courseId.ToString());

        await Assert.ThrowsAnyAsync<EventStoreException>(async () =>
            await store.AppendAsync(newEvent, staleModel.AppendCondition));
    }

    [Fact]
    public async Task BuildDecisionModelAsync_FreshCondition_AppendSucceedsAsync()
    {
        // Arrange: append initial event, build a FRESH decision model
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var freshModel = await store.BuildDecisionModelAsync(EnrollmentCountProjection(courseId));

        // Act: no concurrent write — condition should pass
        var newEvent = new StudentEnrolledEvent(courseId, Guid.NewGuid())
            .ToDomainEvent()
            .WithTag("courseId", courseId.ToString());

        var exception = await Record.ExceptionAsync(async () =>
            await store.AppendAsync(newEvent, freshModel.AppendCondition));

        Assert.Null(exception);
    }

    [Fact]
    public async Task BuildDecisionModelAsync_UnrelatedConcurrentWrite_DoesNotInvalidateConditionAsync()
    {
        // The DCB boundary is scoped — unrelated events must not cause conflicts
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();
        var otherCourseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var model = await store.BuildDecisionModelAsync(EnrollmentCountProjection(courseId));

        // Concurrent write to a completely different course
        await store.AppendAsync(
            new StudentEnrolledEvent(otherCourseId, Guid.NewGuid())
                .ToDomainEvent()
                .WithTag("courseId", otherCourseId.ToString()));

        // Our condition covers courseId only — the write above must not invalidate it
        var newEvent = new StudentEnrolledEvent(courseId, Guid.NewGuid())
            .ToDomainEvent()
            .WithTag("courseId", courseId.ToString());

        var exception = await Record.ExceptionAsync(async () =>
            await store.AppendAsync(newEvent, model.AppendCondition));

        Assert.Null(exception);
    }

    #endregion

    #region Cancellation

    [Fact]
    public async Task BuildDecisionModelAsync_CancelledToken_ThrowsOperationCanceledExceptionAsync()
    {
        var store = CreateEventStore();
        var projection = CourseExistsProjection(Guid.NewGuid());

        using var cts = new CancellationTokenSource();
        cts.Cancel();

        await Assert.ThrowsAnyAsync<OperationCanceledException>(async () =>
            await store.BuildDecisionModelAsync(projection, cts.Token));
    }

    #endregion

    #region Guard clauses

    [Fact]
    public async Task BuildDecisionModelAsync_NullEventStore_ThrowsArgumentNullExceptionAsync()
    {
        var projection = CourseExistsProjection(Guid.NewGuid());

        await Assert.ThrowsAsync<ArgumentNullException>(async () =>
            await ((IEventStore)null!).BuildDecisionModelAsync(projection));
    }

    [Fact]
    public async Task BuildDecisionModelAsync_NullProjection_ThrowsArgumentNullExceptionAsync()
    {
        var store = CreateEventStore();

        await Assert.ThrowsAsync<ArgumentNullException>(async () =>
            await store.BuildDecisionModelAsync((IDecisionProjection<bool>)null!));
    }

    #endregion

    #region Composed 2-projection overload

    [Fact]
    public async Task Compose2_BothStatesReflectAppendedEventsAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();
        var studentId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        await store.AppendAsync(
            new StudentEnrolledEvent(courseId, studentId)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var (courseExists, enrollmentCount, _) = await store.BuildDecisionModelAsync(
            CourseExistsProjection(courseId),
            EnrollmentCountProjection(courseId));

        Assert.True(courseExists);
        Assert.Equal(1, enrollmentCount);
    }

    [Fact]
    public async Task Compose2_SingleReadForBothProjections_ConditionSpansBothQueriesAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();
        var studentId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var (_, _, condition) = await store.BuildDecisionModelAsync(
            CourseExistsProjection(courseId),
            EnrollmentCountProjection(courseId));

        Assert.NotNull(condition.AfterSequencePosition);
        Assert.NotNull(condition.FailIfEventsMatch);
        // Union query must have items from both sub-queries
        Assert.True(condition.FailIfEventsMatch.QueryItems.Count >= 2);
    }

    [Fact]
    public async Task Compose2_StaleCondition_AppendThrowsAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 1)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var (_, _, staleCondition) = await store.BuildDecisionModelAsync(
            CourseExistsProjection(courseId),
            EnrollmentCountProjection(courseId));

        // Concurrent write against one of the queries
        await store.AppendAsync(
            new StudentEnrolledEvent(courseId, Guid.NewGuid())
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var newEvent = new StudentEnrolledEvent(courseId, Guid.NewGuid())
            .ToDomainEvent()
            .WithTag("courseId", courseId.ToString());

        await Assert.ThrowsAnyAsync<EventStoreException>(async () =>
            await store.AppendAsync(newEvent, staleCondition));
    }

    [Fact]
    public async Task Compose2_UnrelatedConcurrentWrite_DoesNotInvalidateConditionAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();
        var otherCourseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        var (_, _, condition) = await store.BuildDecisionModelAsync(
            CourseExistsProjection(courseId),
            EnrollmentCountProjection(courseId));

        // Unrelated write — different courseId, outside our composed query scope
        await store.AppendAsync(
            new StudentEnrolledEvent(otherCourseId, Guid.NewGuid())
                .ToDomainEvent()
                .WithTag("courseId", otherCourseId.ToString()));

        var newEvent = new StudentEnrolledEvent(courseId, Guid.NewGuid())
            .ToDomainEvent()
            .WithTag("courseId", courseId.ToString());

        var exception = await Record.ExceptionAsync(async () =>
            await store.AppendAsync(newEvent, condition));

        Assert.Null(exception);
    }

    #endregion

    #region Composed 3-projection overload

    [Fact]
    public async Task Compose3_AllThreeStatesReflectAppendedEventsAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();
        var studentId = Guid.NewGuid();

        var studentExistsProjection = new DecisionProjection<bool>(
            initialState: false,
            query: Query.FromItems(new QueryItem
            {
                EventTypes = [nameof(StudentEnrolledEvent)],
                Tags = [new Tag("studentId", studentId.ToString())]
            }),
            apply: (_, evt) => evt.Event.Event is StudentEnrolledEvent);

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        await store.AppendAsync(
            new StudentEnrolledEvent(courseId, studentId)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString())
                .WithTag("studentId", studentId.ToString()));

        var (courseExists, enrollmentCount, studentEnrolled, _) = await store.BuildDecisionModelAsync(
            CourseExistsProjection(courseId),
            EnrollmentCountProjection(courseId),
            studentExistsProjection);

        Assert.True(courseExists);
        Assert.Equal(1, enrollmentCount);
        Assert.True(studentEnrolled);
    }

    #endregion

    #region N-ary list overload

    [Fact]
    public async Task NaryBuildDecisionModelAsync_EmptyList_ThrowsArgumentExceptionAsync()
    {
        var store = CreateEventStore();
        IReadOnlyList<IDecisionProjection<bool>> projections = [];

        await Assert.ThrowsAsync<ArgumentException>(async () =>
            await store.BuildDecisionModelAsync(projections));
    }

    [Fact]
    public async Task NaryBuildDecisionModelAsync_NullList_ThrowsArgumentNullExceptionAsync()
    {
        var store = CreateEventStore();

        await Assert.ThrowsAsync<ArgumentNullException>(async () =>
            await store.BuildDecisionModelAsync((IReadOnlyList<IDecisionProjection<bool>>)null!));
    }

    [Fact]
    public async Task NaryBuildDecisionModelAsync_SingleProjection_StateMatchesSingleOverloadAsync()
    {
        var store = CreateEventStore();
        var courseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId.ToString()));

        IReadOnlyList<IDecisionProjection<bool>> projections = [CourseExistsProjection(courseId)];
        var (states, _) = await store.BuildDecisionModelAsync(projections);

        Assert.Single(states);
        Assert.True(states[0]);
    }

    [Fact]
    public async Task NaryBuildDecisionModelAsync_MultipleProjections_EachHasIndependentStateAsync()
    {
        var store = CreateEventStore();
        var courseId1 = Guid.NewGuid();
        var courseId2 = Guid.NewGuid();

        // Append events for course 1 only
        await store.AppendAsync(
            new CourseCreatedEvent(courseId1, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId1.ToString()));

        IReadOnlyList<IDecisionProjection<bool>> projections =
        [
            CourseExistsProjection(courseId1),
            CourseExistsProjection(courseId2)
        ];

        var (states, _) = await store.BuildDecisionModelAsync(projections);

        Assert.Equal(2, states.Count);
        Assert.True(states[0]);   // course 1 exists
        Assert.False(states[1]);  // course 2 does not exist
    }

    [Fact]
    public async Task NaryBuildDecisionModelAsync_ThreeProjections_AllStatesCorrectAsync()
    {
        var store = CreateEventStore();
        var courseId1 = Guid.NewGuid();
        var courseId2 = Guid.NewGuid();
        var courseId3 = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId1, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId1.ToString()));

        await store.AppendAsync(
            new CourseCreatedEvent(courseId2, 10)
                .ToDomainEvent()
                .WithTag("courseId", courseId2.ToString()));

        // course 3 not created

        IReadOnlyList<IDecisionProjection<bool>> projections =
        [
            CourseExistsProjection(courseId1),
            CourseExistsProjection(courseId2),
            CourseExistsProjection(courseId3)
        ];

        var (states, condition) = await store.BuildDecisionModelAsync(projections);

        Assert.Equal(3, states.Count);
        Assert.True(states[0]);
        Assert.True(states[1]);
        Assert.False(states[2]);
        Assert.Equal(2, condition.AfterSequencePosition);
    }

    [Fact]
    public async Task NaryBuildDecisionModelAsync_StaleCondition_AppendThrowsAsync()
    {
        var store = CreateEventStore();
        var courseId1 = Guid.NewGuid();
        var courseId2 = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId1, 1)
                .ToDomainEvent()
                .WithTag("courseId", courseId1.ToString()));

        IReadOnlyList<IDecisionProjection<bool>> projections =
        [
            CourseExistsProjection(courseId1),
            CourseExistsProjection(courseId2)
        ];

        var (_, staleCondition) = await store.BuildDecisionModelAsync(projections);

        // Concurrent write matching one of the projection queries
        await store.AppendAsync(
            new CourseCreatedEvent(courseId2, 20)
                .ToDomainEvent()
                .WithTag("courseId", courseId2.ToString()));

        var newEvent = new StudentEnrolledEvent(courseId1, Guid.NewGuid())
            .ToDomainEvent()
            .WithTag("courseId", courseId1.ToString());

        await Assert.ThrowsAnyAsync<EventStoreException>(async () =>
            await store.AppendAsync(newEvent, staleCondition));
    }

    [Fact]
    public async Task NaryBuildDecisionModelAsync_UnrelatedConcurrentWrite_DoesNotInvalidateConditionAsync()
    {
        var store = CreateEventStore();
        var courseId1 = Guid.NewGuid();
        var courseId2 = Guid.NewGuid();
        var unrelatedCourseId = Guid.NewGuid();

        await store.AppendAsync(
            new CourseCreatedEvent(courseId1, 30)
                .ToDomainEvent()
                .WithTag("courseId", courseId1.ToString()));

        IReadOnlyList<IDecisionProjection<bool>> projections =
        [
            CourseExistsProjection(courseId1),
            CourseExistsProjection(courseId2)
        ];

        var (_, condition) = await store.BuildDecisionModelAsync(projections);

        // Unrelated write — different courseId, not covered by either projection's query
        await store.AppendAsync(
            new CourseCreatedEvent(unrelatedCourseId, 5)
                .ToDomainEvent()
                .WithTag("courseId", unrelatedCourseId.ToString()));

        var newEvent = new StudentEnrolledEvent(courseId1, Guid.NewGuid())
            .ToDomainEvent()
            .WithTag("courseId", courseId1.ToString());

        var exception = await Record.ExceptionAsync(async () =>
            await store.AppendAsync(newEvent, condition));

        Assert.Null(exception);
    }

    #endregion
}

File: tests/build_decision_model_integration_tests.rs
mod common;

use common::{FakeClock, InMemoryStorage};
use marmosa::{
    decision_model::{DecisionModelExt, DecisionProjection},
    domain::{DomainEvent, EventData, EventRecord, Query, QueryItem, Tag},
    event_store::{EventStore, OpossumStore},
};
use std::sync::Arc;

// Minimal event types for tests
#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct CourseCreatedEvent {
    course_id: String,
    max_students: i32,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct StudentEnrolledEvent {
    course_id: String,
    student_id: String,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct UnrelatedEvent {
    data: String,
}

// Helper projections
struct CourseExistsProjection {
    query: Query,
}

impl CourseExistsProjection {
    fn new(course_id: impl Into<String>) -> Self {
        let query = Query {
            items: vec![QueryItem {
                event_types: vec!["CourseCreatedEvent".to_string()],
                tags: vec![Tag {
                    key: "course_id".to_string(),
                    value: course_id.into(),
                }],
            }],
        };
        Self { query }
    }
}

impl DecisionProjection for CourseExistsProjection {
    type State = bool;

    fn initial_state(&self) -> Self::State {
        false
    }

    fn query(&self) -> &Query {
        &self.query
    }

    fn apply(&self, _state: Self::State, evt: &EventRecord) -> Self::State {
        evt.event.event_type == "CourseCreatedEvent"
    }
}

struct EnrollmentCountProjection {
    query: Query,
}

impl EnrollmentCountProjection {
    fn new(course_id: impl Into<String>) -> Self {
        let query = Query {
            items: vec![QueryItem {
                event_types: vec!["StudentEnrolledEvent".to_string()],
                tags: vec![Tag {
                    key: "course_id".to_string(),
                    value: course_id.into(),
                }],
            }],
        };
        Self { query }
    }
}

impl DecisionProjection for EnrollmentCountProjection {
    type State = i32;

    fn initial_state(&self) -> Self::State {
        0
    }

    fn query(&self) -> &Query {
        &self.query
    }

    fn apply(&self, count: Self::State, evt: &EventRecord) -> Self::State {
        if evt.event.event_type == "StudentEnrolledEvent" {
            count + 1
        } else {
            count
        }
    }
}

fn create_event_store() -> Arc<OpossumStore<Arc<InMemoryStorage>, FakeClock>> {
    let storage = Arc::new(InMemoryStorage::new());
    Arc::new(OpossumStore::new(storage, FakeClock::new(1000)))
}

#[tokio::test]
async fn build_decision_model_async_empty_store_state_is_initial_state_async() {
    let store = create_event_store();
    let course_id = uuid::Uuid::new_v4().to_string();

    let model = store
        .build_decision_model_async(CourseExistsProjection::new(course_id.clone()))
        .await
        .unwrap();

    assert_eq!(model.state, false);
}

#[tokio::test]
async fn build_decision_model_async_empty_store_after_sequence_position_is_null_async() {
    let store = create_event_store();
    let course_id = uuid::Uuid::new_v4().to_string();

    let model = store
        .build_decision_model_async(CourseExistsProjection::new(course_id.clone()))
        .await
        .unwrap();

    assert_eq!(model.append_condition.after_sequence_position, None);
}

#[tokio::test]
async fn build_decision_model_async_empty_store_fail_if_events_match_is_projection_query_async() {
    let store = create_event_store();
    let course_id = uuid::Uuid::new_v4().to_string();
    let projection = CourseExistsProjection::new(course_id.clone());

    let model = store
        .build_decision_model_async(CourseExistsProjection::new(course_id.clone()))
        .await
        .unwrap();

    // The method return type for append condition contains `fail_if_events_match` of type Option<Query>
    assert_eq!(
        model.append_condition.fail_if_events_match,
        projection.query().clone()
    );
}

#[tokio::test]
async fn build_decision_model_async_after_append_state_reflects_events_async() {
    let store = create_event_store();
    let course_id = uuid::Uuid::new_v4().to_string();

    store
        .append_async(
            vec![EventData {
                event_id: uuid::Uuid::new_v4().to_string(),
                metadata: None,
                event: DomainEvent {
                    event_type: "CourseCreatedEvent".to_string(),
                    data: serde_json::to_string(&CourseCreatedEvent {
                        course_id: course_id.clone(),
                        max_students: 10,
                    })
                    .unwrap(),
                    tags: vec![Tag {
                        key: "course_id".to_string(),
                        value: course_id.clone(),
                    }],
                },
            }],
            None,
        )
        .await
        .unwrap();

    let model = store
        .build_decision_model_async(CourseExistsProjection::new(course_id.clone()))
        .await
        .unwrap();

    assert_eq!(model.state, true);
}

#[tokio::test]
async fn build_decision_model_async_after_multiple_enrollments_count_is_correct_async() {
    let store = create_event_store();
    let course_id = uuid::Uuid::new_v4().to_string();

    for i in 0..3 {
        store
            .append_async(
                vec![EventData {
                    event_id: uuid::Uuid::new_v4().to_string(),
                    metadata: None,
                    event: DomainEvent {
                        event_type: "StudentEnrolledEvent".to_string(),
                        data: serde_json::to_string(&StudentEnrolledEvent {
                            course_id: course_id.clone(),
                            student_id: format!("student-{}", i),
                        })
                        .unwrap(),
                        tags: vec![Tag {
                            key: "course_id".to_string(),
                            value: course_id.clone(),
                        }],
                    },
                }],
                None,
            )
            .await
            .unwrap();
    }

    let model = store
        .build_decision_model_async(EnrollmentCountProjection::new(course_id.clone()))
        .await
        .unwrap();

    assert_eq!(model.state, 3);
}

#[tokio::test]
async fn build_decision_model_nary_async_three_projections_all_states_correct_async() {
    let store = create_event_store();
    let course_id = uuid::Uuid::new_v4().to_string();

    store
        .append_async(
            vec![
                EventData {
                    event_id: uuid::Uuid::new_v4().to_string(),
                    metadata: None,
                    event: DomainEvent {
                        event_type: "CourseCreatedEvent".to_string(),
                        data: serde_json::to_string(&CourseCreatedEvent {
                            course_id: course_id.clone(),
                            max_students: 10,
                        })
                        .unwrap(),
                        tags: vec![Tag {
                            key: "course_id".to_string(),
                            value: course_id.clone(),
                        }],
                    },
                },
                EventData {
                    event_id: uuid::Uuid::new_v4().to_string(),
                    metadata: None,
                    event: DomainEvent {
                        event_type: "StudentEnrolledEvent".to_string(),
                        data: serde_json::to_string(&StudentEnrolledEvent {
                            course_id: course_id.clone(),
                            student_id: "student-1".to_string(),
                        })
                        .unwrap(),
                        tags: vec![Tag {
                            key: "course_id".to_string(),
                            value: course_id.clone(),
                        }],
                    },
                },
                EventData {
                    event_id: uuid::Uuid::new_v4().to_string(),
                    metadata: None,
                    event: DomainEvent {
                        event_type: "StudentEnrolledEvent".to_string(),
                        data: serde_json::to_string(&StudentEnrolledEvent {
                            course_id: course_id.clone(),
                            student_id: "student-2".to_string(),
                        })
                        .unwrap(),
                        tags: vec![Tag {
                            key: "course_id".to_string(),
                            value: course_id.clone(),
                        }],
                    },
                },
            ],
            None,
        )
        .await
        .unwrap();

    let p1 = EnrollmentCountProjection::new(course_id.clone());
    let p2 = EnrollmentCountProjection::new(course_id.clone());
    let p3 = EnrollmentCountProjection::new(course_id.clone());
    let projections: Vec<&(dyn DecisionProjection<State = i32> + Send + Sync)> =
        vec![&p1, &p2, &p3];

    let (states, condition) = store
        .build_decision_model_nary_async(&projections)
        .await
        .unwrap();

    assert_eq!(states, vec![2, 2, 2]);
    assert!(condition.after_sequence_position.is_some());
}
